Source/NameServer/SBase.cpp

```diff
@@ -0,0 +1,1550 @@
+   {
+      Post(710, "QueryByAttribute", int2str(Urank));
+      return AddressBook::ERR_INVALID_TASK; 
+   }
+   unsigned err = SUCCESS;
+   // query asks for attributes by name (should this be by attribute ID for
+   // compactness?)
+   vector<string> attrs; 
+   msg->GetX(2,attrs);              // third dynamic field is the attribute set
+   int srcProc = msg->Src();        // need the querying source to send a reply
+   msg->comm = Comms[comm];
+   msg->Src(msg->Tgt());            // build the reply message
+   msg->L(1,Q::RPLY);
+   vector<SymAddr_t> devAddrs;
+   vector<string> devNames;
+   vector<string> devAttrs;
+   // now grab all the devices having one or more of the attributes asked for
+   WALKVECTOR(string, attrs, a)     
+   {
+      const RecordVect_t* devRecords;
+      if ((err = FindByAttribute(taskName, *a, devRecords)))
+      {
+	  // filter out nonexistent task/nonexistent device queries. An
+	  // ERR_INVALID_MAP indicates the task has a dirty device table, and
+	  // should be rebuilt before asking for devices.
+	  if (err == ERR_TASK_NOT_FOUND || err == ERR_INVALID_MAP)
+	  {
+	     // Q::TNF => task not found; Q::NF => no matching records found
+             msg->L(3, err == ERR_TASK_NOT_FOUND ? Q::TNF : Q::NF); 
+             msg->Send(srcProc); // send not found response
+             return SUCCESS;
+	  }
+	  continue; // no devices with this attribute were found
+      }
+      devAttrs.push_back(*a); // send the actual attributes matched in the reply
+      for (vector<const Record_t*>::const_iterator device = devRecords->begin();
+	   device != devRecords->end(); device++)
+      {
+	  devAddrs.push_back((*device)->Address);
+	  devNames.push_back((*device)->Name);
+      }
+   }	
+   if (!devAddrs.size()) msg->L(3, Q::NF); // no attributes matched
+   else
+   {
+      msg->Put<SymAddr_t>(0,&devAddrs); // otherwise reply with the matching 
+      msg->PutX(1,&devNames);           // devices' name-address pairs
+      msg->PutX(2,&devAttrs);
+   }
+   msg->Send(srcProc);
+   return SUCCESS;
+}
+
+//==============================================================================
+/*
+  QueryDevIAll - Gets all devices for a task. We return every device in a single
+  PMsg_p. This could get VERY big, but since there are no certainties the
+  receiving device knows how many devices it's expecting, chunking the reply
+  into bite-sized messages makes it impossible for the receiver to know when it
+  can stop receiving. An alternative would be to insert the device count in the
+  message, along with possibly a sequence number, but for a first revision this
+  is just adding complication for possibly little benefit. To be tested.
+*/
+//==============================================================================
+unsigned SBase::QueryDevIAll(PMsg_p *msg, unsigned comm)
+{
+   string taskName = msg->Zname(0); // as usual, first static field is the task
```

Comment:
DRY

Reply:
The whole thing is structured so that many of the fields in the message contain the same information retrieved in the same way. But I am extracting them in each command-response directly so that when looking through the code you don't have to go hunting through multiple functions for each little bit of extraction - again, to avoid over-atomisation.

Reply:
> The whole thing is structured so that many of the fields in the message contain the same information retrieved in the same way. 

So why not abstract it? By not abstracting it, you're brittle to any potential change in the structure later.

> But I am extracting them in each command-response directly so that when looking through the code you don't have to go hunting through multiple functions for each little bit of extraction

If you read these query methods in sequence, and if you call the same method each time, it's more clear to a reader that you are doing the same thing. The reader looks at the function once (at most) and, if the method is named appropriately, you won't need to look at the definition to understand what it does.

> hunting through multiple functions for each little bit of extraction

It's trivial to lookup a method in this way - the call stack here is pretty flat, and the function would be in the same file. 

We're still several levels away from "over-atomisation" in my opinion. I argue that since you reduce code size, duplication, and improve clarity and modularity at the cost of one stack level, the tradeoff of abstraction is certainly worth it here.

"*In my opinion*" being the key clause here.

Reply:
"So why not abstract it? By not abstracting it, you're brittle to any potential change in the structure later."

Partly because the fact that several fields actually contain the same data is a design convenience as opposed to an interface specification - it is the intent that each of the commands could potentially change their interface independently. If you create an abstration, then commands become co-dependent on the same argument set.

The other reason is that in fact, whilst many commands do share identical arguments, the subset of shared arguments varies, so that it's not easy to define an abstraction that conveniently captures a large number of the command subset completely. About the only truly common argument is the task name - and even that isn't shared in some commands.

Let's imagine, for example, that you have 8 commands containing 4 arguments apiece. Call these arguments A, B, C, and D. Also imagine that in many cases these arguments have identical values, A*, B*, C*, and D*. So now what is the right abstraction if you have a command structure like this:

1 (A*,B*,C*,D)
2 (A*,B*,C,D*)
3 (A*,B,C*,D*)
4 (A*,B,C,D*)
5 (A,B*,C*,D*)
6 (A,B*,C,D*)
7 (A,B*,C*,D)
8 (A,B,C*,D*)

?

As you can see (and this is the situation we actually have), lots of arguments are shared, but
there is no subset that you can pick that's universal. Everything is permuted, so there's a lot of repetition but strangely little commonality. 

If you do try to abstract bits, you just end up with a lot of little abstract methods that 
have to be mixed combinatorially to generate the result for any given command, which doesn't lead to easily readable code and doesn't substantially reduce code size either.